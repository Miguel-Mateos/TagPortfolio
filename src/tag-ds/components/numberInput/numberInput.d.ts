import React from 'react';
export interface INumberInputProps {
    disabled?: boolean;
    className?: string;
    error?: boolean;
    required?: boolean;
    defaultValue?: number;
    onChange?: (num: number) => void;
    boundaries?: {
        max: number;
        min: number;
    };
    placeholder?: {
        icon: React.ReactElement;
        text: string;
    };
    label?: string;
    [others: string]: any;
}
export declare const NumberInput: React.FC<INumberInputProps>;
