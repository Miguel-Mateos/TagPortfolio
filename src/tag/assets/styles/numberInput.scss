.tag-ds {

  .number-input, &.number-input {

    input[type=number] {
      text-align: center;
    }
    
    input[type=number]::-webkit-inner-spin-button,
    input[type=number]::-webkit-outer-spin-button
    {
      -webkit-appearance: none;
    }
    
    .input-number-container {
      position: relative;
    }
    
    label ~ .input-number-container {
      margin-top: var(--unit);
    }
    
    button.input-number-minus {
      @extend %button-icon_interactive;
      padding: 0;
      position: absolute;
      height: calc(3 * var(--unit));
      width: calc(3 * var(--unit));
      left: var(--unit);
      background-color: var(--neutral100);
      display: flex;
      justify-content: center;
      align-items: center;
      z-index: 1;
      border-radius: 50%;
      transform: translateY(50%);
      bottom: 50%;
      font-weight: 500;

    }
    
    button.input-number-plus{
      @extend %button-icon_interactive;
      padding: 0;
      position: absolute;
      font-weight: 500;
      height: calc(3 * var(--unit));
      width: calc(3 * var(--unit));
      right: var(--unit);
      background-color: var(--neutral100);
      display: flex;
      justify-content: center;
      align-items: center;
      z-index: 1;
      border-radius: 50%;
      transform: translateY(-50%);
      top: 50%;
    }
  }
}