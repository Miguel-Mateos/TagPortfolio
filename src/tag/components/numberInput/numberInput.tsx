import React, {  useState } from 'react';

const testId = (base: string|undefined|null, extension: string) => base ? base+'-'+extension : undefined; 

export interface INumberInputProps {
  disabled?: boolean,
  className?: string
  error?: boolean
  required?: boolean
  defaultValue?: number
  onChange?: (num: number) => void
  boundaries?: {
    max: number
    min: number
  },
  placeholder?: {
    icon: React.ReactElement
    text: string
  }
  label?: string
  [others: string]: any
}

export const NumberInput: React.FC<INumberInputProps> = ({
  disabled,
  className,
  error,
  required,
  defaultValue,
  onChange,
  boundaries,
  placeholder,
  label,
  ...rest
}) => {
  const [value, setValue] = useState(defaultValue || '0');

  const handleIncrease = () => {
    setValue(prev =>  !handleBoundaries(+prev+1) ? (+prev + 1).toString() : prev);
    if (onChange) onChange(+value + 1);
  };
  const handleDecrease = () => {
    setValue(prev => !handleBoundaries(+prev-1) ? (+prev - 1).toString() : prev);
    if (onChange) onChange(+value - 1);
  };
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const tValue = e?.currentTarget?.value;
    setValue(tValue);
    if (handleBoundaries(+tValue) !== 'error') {
      if (onChange) onChange(+tValue);
    }
  };

  const handleBoundaries = (val: number) => {
    if (boundaries) {
      if (boundaries.max < val || boundaries.min > val) return 'error';
      return '';
    }
  }; 

  return (
    <div className={`number-input input-wrapper${disabled ? '_disabled' : ''} ${className || ''}
      ${(error || handleBoundaries(+value)) ? 'error' : ''}`}
    data-testid={rest['data-testid'] ? rest['data-testid'] : undefined}
    >
      {label && <label className='caption'>{label}{required && <small>*</small>}</label>}
      <div className='input-number-container'>
        <button disabled={disabled} data-testid={testId(rest['data-testid'], 'decrease')}
          onClick={handleDecrease} className='input-number-minus'>-</button>
        <input
          value={value}
          onChange={handleChange}
          type='number'
          className='input-container'
          disabled={disabled}
          placeholder='hola'
          data-testid={testId(rest['data-testid'], 'input')}
        />
        <button disabled={disabled} data-testid={testId(rest['data-testid'], 'increase')}
          onClick={handleIncrease} className='input-number-plus'>+</button>
      </div>
      {placeholder && <p className='input-helper-text'>
        {placeholder.icon}
        {placeholder.text}
      </p>}
    </div>
  );
};