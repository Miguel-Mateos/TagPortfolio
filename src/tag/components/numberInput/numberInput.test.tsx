import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { INumberInputProps, NumberInput } from './numberInput';
import userEvent from '@testing-library/user-event';

const TestNumber = (args: INumberInputProps) => {
  return <NumberInput {...args}/>;
};

test('render Number Input and match snapshot', () => {
  const component = renderer.create(<TestNumber />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('increment Number input value', () => {
  const { getByTestId } = render(<TestNumber data-testid='number-input'/>);
  const increment = getByTestId('number-input-increase');
  const value = getByTestId('number-input-input');
  fireEvent.click(increment);
  expect(value).toHaveValue(1);
});

test('decrement Number input value', () => {
  const { getByTestId } = render(<TestNumber defaultValue={1} data-testid='number-input'/>);
  const decrement = getByTestId('number-input-decrease');
  const value = getByTestId('number-input-input');
  fireEvent.click(decrement);
  expect(value).toHaveValue(0);
});

test('change value input value with keyboard', () => {
  const { getByTestId } = render(<TestNumber defaultValue={1} data-testid='number-input'/>);
  const input = getByTestId('number-input-input');
  userEvent.clear(input);
  userEvent.type(input, '23');
});

test('surpass boundary by max', () => {
  const { getByTestId } = render(<TestNumber defaultValue={1} boundaries={{max: 10, min: 0 }} data-testid='number-input'/>);
  const component = getByTestId('number-input');
  const input = getByTestId('number-input-input');
  userEvent.clear(input);
  userEvent.type(input, '23');
  expect(component).toHaveClass('error');
});

test('surpass boundary by min', () => {
  const { getByTestId } = render(<TestNumber defaultValue={1} boundaries={{max: 10, min: 0 }} data-testid='number-input'/>);
  const component = getByTestId('number-input');
  const input = getByTestId('number-input-input');
  userEvent.clear(input);
  userEvent.type(input, '-1');
  expect(component).toHaveClass('error');
});

describe('call and render onChange properly', () => {
  test ('witch increase', () => {
    const mock = jest.fn();
    const { getByTestId } = render(<TestNumber onChange={mock} data-testid='number-input'/>);
    const increment = getByTestId('number-input-increase');
    userEvent.click(increment);
    expect(mock).toBeCalledTimes(1);
  });

  test ('witch decrease', () => {
    const mock = jest.fn(); 
    const { getByTestId } = render(<TestNumber onChange={mock} data-testid='number-input'/>);
    const decrement = getByTestId('number-input-decrease');
    userEvent.click(decrement);
    expect(mock).toBeCalledTimes(1);
  });

  test ('with keyboard', () => {
    const mock = jest.fn();
    const { getByTestId } = render(<TestNumber onChange={mock} data-testid='number-input'/>);
    const input = getByTestId('number-input-input');
    userEvent.type(input, '2');
    expect(mock).toBeCalledTimes(1);
  });
});

test('disabled Number Input', () => {
  const { getByTestId } = render(<TestNumber disabled data-testid='number-input'/>);
  const component = getByTestId('number-input');
  expect(component).toHaveClass('input-wrapper_disabled');
});