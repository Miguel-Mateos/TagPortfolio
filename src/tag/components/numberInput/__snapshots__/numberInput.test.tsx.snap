// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Number Input and match snapshot 1`] = `
<div
  className="number-input input-wrapper 
      "
>
  <div
    className="input-number-container"
  >
    <button
      className="input-number-minus"
      onClick={[Function]}
    >
      -
    </button>
    <input
      className="input-container"
      onChange={[Function]}
      placeholder="hola"
      type="number"
      value="0"
    />
    <button
      className="input-number-plus"
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;
